kernel/src/net/socket/netlink/multicast_group.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use super::addr::{GroupId, PortNum};
+use crate::prelude::*;
+
+/// A netlink multicast group.
+///
+/// A group can contain multiple sockets,
+/// each identified by its bound port number.
+pub struct MulticastGroup {
+    members: BTreeSet<PortNum>,
+}
+
+impl MulticastGroup {
+    /// Creates a new multicast group.
+    pub const fn new() -> Self {
+        Self {
+            members: BTreeSet::new(),
+        }
+    }
+
+    /// Returns whether the group contains a member.
+    #[expect(unused)]
+    pub fn contains_member(&self, port_num: PortNum) -> bool {
+        self.members.contains(&port_num)
+    }
+
+    /// Adds a new member to the multicast group.
+    pub fn add_member(&mut self, port_num: PortNum) {
+        debug_assert!(!self.members.contains(&port_num));
+        self.members.insert(port_num);
+    }
+
+    /// Removes a member from the multicast group.
+    pub fn remove_member(&mut self, port_num: PortNum) {
+        debug_assert!(self.members.contains(&port_num));
+        self.members.remove(&port_num);
+    }
+}
+
+/// A set of group IDs.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct GroupIdSet(u32);
+
+impl GroupIdSet {
+    /// Creates a new empty `GroupIdSet`.
+    pub const fn new_empty() -> Self {
+        Self(0)
+    }
+
+    /// Creates a new `GroupIdSet` with multiple groups.
+    ///
+    /// Each 1 bit in `groups` represent a group.
+    pub const fn new(groups: u32) -> Self {
+        Self(groups)
+    }
+
+    /// Creates an iterator over all group IDs.
+    pub const fn ids_iter(&self) -> GroupIdIter<'_> {
+        GroupIdIter::new(self)
+    }
+
+    /// Adds a new group.
+    ///
+    /// If the group already exists, this method will return an error.
+    pub fn add_group(&mut self, group_id: GroupId) -> Result<()> {
+        if group_id >= 32 {
+            return_errno_with_message!(Errno::EINVAL, "invalid group ID");
+        }
+
+        let mask = 1u32 << group_id;
+        if self.0 & mask != 0 {
+            return_errno_with_message!(Errno::EINVAL, "group id already exists");
+        }
+        self.0 |= mask;
+
+        Ok(())
+    }
+
+    /// Sets new groups.
+    pub fn set_groups(&mut self, new_groups: u32) {
+        self.0 = new_groups;
+    }
+
+    /// Clears all groups.
+    pub fn clear(&mut self) {
+        self.0 = 0;
+    }
+
+    /// Checks if the set of group IDs is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+
+    /// Returns the group IDs as a u32.
+    pub fn as_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Iterator over a set of group IDs.
+pub struct GroupIdIter<'a> {
+    groups: &'a GroupIdSet,
```

Comment:
We can store a `groups: u32`, and use [`trailing_zeros`](https://doc.rust-lang.org/std/primitive.u32.html#method.trailing_zeros) to find the lowest one bit.